Gulp serve only serve either compiled or uncompiled version

zhouyx opened the pull request:
For #7581 
Please let me know if the approach looks good in general.  
Will get to integration test change in different PR.  Also it would require some doc change. 
Heroku would by default run max version. Can someone point to me where the config is for heroku?

zhouyx commented on src/service/extension-location.js:
```diff
@@ -59,32 +47,10 @@ export function calculateExtensionScriptUrl(location, extensionId, isLocalDev,
-export function calculateEntryPointScriptUrl(location, entryPoint, isLocalDev,
-    isTest) {
-  const base = calculateScriptBaseUrl(location, isLocalDev, isTest);
-  const serveMax = isLocalDev && isMax(location);
```
@choumx @jridgewell any ideas why we don't take `isTest` into consideration here?

choumx replied:
Probably just an oversight (or deemed unnecessary). We'd also need `isUsingCompiledJs` to make sure we test min builds with `--compiled` here.

zhouyx replied:
good news is with this PR, gulp server will use min files with `--compiled` flag.  we don't need to deal with `isUsingCompiledJs` anymore.

cramforce commented on build-system/tasks/serve.js:
```diff
@@ -23,6 +23,18 @@ var nodemon = require('nodemon');
 function serve() {
+  // Get the serve mode
+  if (argv.compiled) {
+    process.env.SERVE_MODE = 'min';
```
Could we make this changeable at runtime?

zhouyx replied:
I can add the check for `app.use('/serve_mode=?')` to change it at runtime. If this is what you mean?

cramforce commented:
Yep.

On Tue, Apr 4, 2017 at 10:33 AM, Yuxuan Zhou <notifications@github.com>
wrote:

> *@zhouyx* commented on this pull request.
> ------------------------------
>
> In build-system/tasks/serve.js
> <https://github.com/ampproject/amphtml/pull/8559#discussion_r109728432>:
>
> > @@ -23,6 +23,18 @@ var nodemon = require('nodemon');
>   * Starts a simple http server at the repository root
>   */
>  function serve() {
> +  // Get the serve mode
> +  if (argv.compiled) {
> +    process.env.SERVE_MODE = 'min';
>
> I can add the check for app.use('/serve_mode=?') to change it at runtime.
> If this is what you mean?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/8559#discussion_r109728432>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFeT9YiXgAnBk3ph9zgcfPHEC4P_AF3ks5rsn7ogaJpZM4MyDcK>
> .
>

zhouyx commented:
PTAL 👀

zhouyx commented:
Ping~

zhouyx commented:
Friendly ping on this PR. @cramforce Do we still want to keep it? : )

cramforce commented:
I'm just waiting for others to review :)

zhouyx commented:
LOLOL 😄  Ping @erwinmombay @lannka and @dvoytenko

dvoytenko commented on build-system/app.js:
```diff
@@ -28,9 +28,31 @@ var jsdom = require('jsdom');
 var path = require('path');
 var request = require('request');
 var url = require('url');
+var mode = process.env.SERVE_MODE;
```
Can you document here what's default is?

dvoytenko commented on build-system/app.js:
```diff
@@ -28,9 +28,31 @@ var jsdom = require('jsdom');
+    process.env.SERVE_MODE = newMode;
+    info = '<h2>Serve mode changed to ' + newMode + '</h2>';
+  } else {
+    info = '<h2>Serve mode ' + newMode + ' is not support. </h2>';
```
"ed"

dvoytenko commented on build-system/app.js:
```diff
@@ -486,19 +511,31 @@ app.use('/impression-proxy/', function(req, res) {
 // Example:
 // http://localhost:8000/max/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
 app.use('/max/', function(req, res) {
+  if (mode != 'max') {
```
How about we just rename it `/proxy/` or `/cdn/`? Then, we'll just need one and we can remove `/min/`

dvoytenko commented on src/service/extension-location.js:
```diff
@@ -21,14 +21,11 @@ import {getMode} from '../mode';
-    if (isTest || isMax(location) || isMin(location)) {
-      return `${location.protocol}//${location.host}/dist`;
-    }
+    return `${location.protocol}//${location.host}/dist`;
```
I really dislike `isLocalDev` being passed here. At the minimum, it kills the DCE. I suppose we can live with it, at least for this PR. But if it's easy to remove, let's do.

zhouyx replied:
I don't know a good way to remove `isLocalDev` here 😢

lannka replied:
Agree, DCE of localDev code can be a separate project :-). I do see many code can be DCEed if we simplify our isLocalDev logic [here](https://github.com/ampproject/amphtml/blob/master/src/mode.js#L84)

dvoytenko commented on src/service/extension-location.js:
```diff
@@ -59,32 +47,10 @@ export function calculateExtensionScriptUrl(location, extensionId, isLocalDev,
-  return path.indexOf('.min') >= 0 || path.substr(0, 5) == '/min/';
+export function calculateEntryPointScriptUrl(location, entryPoint, isLocalDev) {
+  const base = calculateScriptBaseUrl(location, isLocalDev);
+  // QQQ: why we don't use .max version for test here?
```
Remove?

zhouyx replied:
confirmed. removed.

zhouyx commented:
addressed comments from @dvoytenko. PTAL
@jridgewell would you take a look at the change related to service worker once you're available. 
@bpaduch will need you to review the change on developing.md.
Thanks a lot!

bpaduch commented on contributing/DEVELOPING.md:
```diff
@@ -65,16 +65,17 @@ The Quick Start Guide's  [One-time setup](getting-started-quick.md#one-time-setu
+### Serve Mode
+There are 3 serving modes:
+- MAX mode serves unminified AMP. You want to use this during normal dev. `gulp` serves MAX mode by default.
+- MIN mode serves minified AMP. This is closer to the prod setup. This is only avaialable after running `gulp dist --fortesting`. Serve MIN mode by adding `--compiled` to `gulp` command.
```
correct spelling of "available"

bpaduch commented on contributing/DEVELOPING.md:
```diff
@@ -65,16 +65,17 @@ The Quick Start Guide's  [One-time setup](getting-started-quick.md#one-time-setu
+There are 3 serving modes:
+- MAX mode serves unminified AMP. You want to use this during normal dev. `gulp` serves MAX mode by default.
+- MIN mode serves minified AMP. This is closer to the prod setup. This is only avaialable after running `gulp dist --fortesting`. Serve MIN mode by adding `--compiled` to `gulp` command.
+- CDN mode serves prod. This file would not reflect your local changes. Serve CDN mode by adding `--cdn` to `gulp` command.
```
"this file" - what file?  Please clarify

bpaduch commented on contributing/DEVELOPING.md:
```diff
@@ -65,16 +65,17 @@ The Quick Start Guide's  [One-time setup](getting-started-quick.md#one-time-setu
+- CDN mode serves prod. This file would not reflect your local changes. Serve CDN mode by adding `--cdn` to `gulp` command.
 
-For each example there are 3 modes:
+To switch serving mode in runtime. please to to http://localhost:8000/serve_mode=$mode. $mode can be `max/min/cdn`.
```
Rewrite as:

To switch serving modes during runtime, go to http://localhost:8000/serve_mode=$mode and set the `$mode` to one of the following values: `max`, `min`, or  `cdn`.

bpaduch commented on contributing/DEVELOPING.md:
```diff
@@ -84,15 +85,14 @@ For any public AMP document like: http://output.jsbin.com/pegizoq/quiet
 
-- normal sources: http://localhost:8000/max/output.jsbin.com/pegizoq/quiet
-- minified: http://localhost:8000/min/output.jsbin.com/pegizoq/quiet
+http://localhost:8000/proxy/output.jsbin.com/pegizoq/quiet
```
Correct wording of this section to:

For any public AMP document like http://output.jsbin.com/pegizoq/quiet, you can access it with the local JS at http://localhost:8000/proxy/output.jsbin.com/pegizoq/quiet.

bpaduch commented on contributing/DEVELOPING.md:
```diff
@@ -84,15 +85,14 @@ For any public AMP document like: http://output.jsbin.com/pegizoq/quiet
+http://localhost:8000/proxy/output.jsbin.com/pegizoq/quiet
 
-When accessing `min` urls make sure you run `gulp dist` with the `--fortesting`
+Note local proxy will serve minified or unminified JS based on current serve mode. When serve mode is `cdn` local proxy will serve remote JS.
```
Rewrite as follows:

`**`Note`**`: The local proxy will serve minified or unminified JS based on the current serve mode. When the serve mode is `cdn`, the local proxy will serve remote JS.

zhouyx commented:
modified doc.
Ping @jridgewell to take a look. 😄

lannka commented on build-system/app.js:
```diff
@@ -28,9 +28,32 @@ var jsdom = require('jsdom');
+});
+
+// Deprecate usage of .min.html/.max.html
+app.use(['/examples/*.(min|max).html', '/test/manual/*.(min|max).html',
```
app.get

lannka commented on build-system/app.js:
```diff
@@ -28,9 +28,32 @@ var jsdom = require('jsdom');
 
 app.use(bodyParser.json());
 
+app.use('/serve_mode=:mode', function (req, res, next) {
```
app.get

lannka commented on build-system/app.js:
```diff
@@ -28,9 +28,32 @@ var jsdom = require('jsdom');
+    process.env.SERVE_MODE = newMode;
+    info = '<h2>Serve mode changed to ' + newMode + '</h2>';
+  } else {
+    info = '<h2>Serve mode ' + newMode + ' is not supported. </h2>';
```
nit: `res.status(400).send(info)`

lannka commented on build-system/tasks/serve.js:
```diff
@@ -27,6 +27,18 @@ var useHttps = argv.https != undefined;
+    process.env.SERVE_MODE = 'cdn';
+    util.log(util.colors.green('Serving current prod js'));
+  } else {
+    process.env.SERVE_MODE = 'max';
```
nit: to be consistent with the flags, can we name them: `compiled|cdn|default`?

zhouyx replied:
But I think we are more familiar with `min`/`max`. And it's not obvious right now that `default` is `max` mode.

lannka replied:
But I thought this whole PR is to get rid of min/max concept, and only stay with the "compiled" concept.

lannka commented on build-system/app.js:
```diff
@@ -28,9 +28,32 @@ var jsdom = require('jsdom');
 var request = require('request');
 var url = require('url');
+// SERVE_MODE is defaulted to be 'max' if not specified
+var mode = process.env.SERVE_MODE;
```
how about just `MODE`?

zhouyx replied:
I like `SERVE_MODE` better because we already have `SERVE_PORT` `SERVE_HOST`...

jridgewell commented on build-system/app.js:
```diff
@@ -738,9 +752,53 @@ app.get([ fakeAdNetworkDataDir + '/*', cloudflareDataDir + '/*'], function(req,
+        // Require url from cdn
+        var filePath = 'https://cdn.ampproject.org/' + fileName;
+        request(filePath, function(error, response) {
+          res.send(response);
```
Check for the error.

jridgewell commented on build-system/app.js:
```diff
@@ -738,9 +752,53 @@ app.get([ fakeAdNetworkDataDir + '/*', cloudflareDataDir + '/*'], function(req,
+    // This will not be useful until extension-location.js change in prod
+    // Require url from cdn
+    request(filePath, function (error, response) {
+      res.send(response);
```
Check for the error.

jridgewell commented on build-system/app.js:
```diff
@@ -738,9 +752,53 @@ app.get([ fakeAdNetworkDataDir + '/*', cloudflareDataDir + '/*'], function(req,
+      }
+      if (mode == 'max') {
+        var fileUrl = req.url;
+        req.url = fileUrl.substr(0, fileUrl.length - 3) + '.max.js';
```
Nit: `req.url = req.url.replace(/\.js$/, '.max.js')`

zhouyx commented:
Done. Will finally merge this after travis pass 🤓

muxin commented on test/manual/accessibility/aria-propagation.amp.html:
```diff
@@ -6,9 +6,9 @@
-    <script async custom-element="amp-video" src="../../../dist/v0/amp-video-0.1.max.js"></script>
+    <script async custom-element="amp-anim" src="https://cdn.ampproject.org/v0/amp-anim-0.1.js"></script>
+    <script async custom-element="amp-audio" src="https://cdn.ampproject.org/v0/amp-audio-0.1.js"></script>
+    <script async custom-element="amp-video" src="https://cdn.ampproject.org//amp-video-0.1.js"></script>
```
Missing v0 here?
